resources/net.go

```diff
@@ -0,0 +1,726 @@
+func (obj *NetRes) ifaceCheckApply(apply bool) (checkOK bool, _ error) {
+	var ok bool
+
+	// create a bus object pointing to the adapter
+	busObj := obj.gdbus.Object(DBusNetworkdInterface, obj.IfacePath)
+	// get the object's operational state
+	call := busObj.Call(DBusPropertiesGet, 0, DBusNetworkdLinkInterface, OpState)
+	if call.Err != nil {
+		return false, call.Err
+	}
+	// save the operational state to the struct
+	obj.OpState, ok = call.Body[0].(dbus.Variant)
+	if !ok {
+		return false, fmt.Errorf("error casting opstate")
+	}
+
+	// if the interface's state is correct, we're done
+	if obj.State == IfaceDown && obj.OpState == dbus.MakeVariant(OpStateOff) {
+		return true, nil
+	}
+	if obj.State == IfaceUp {
+		if obj.OpState == dbus.MakeVariant(OpStateRoutable) {
+			return true, nil
+		}
+		// If the connection is lost and timeout is set, retry once a second
+		// until the timeout expires. If the connection comes back, continue
+		// as if nothing happened. Error if the timeout expires (or isn't set.)
+		if obj.OpState == dbus.MakeVariant(OpStateNoCarrier) {
+			for i := 0; i < int(obj.Timeout); i++ {
+				time.Sleep(1 * time.Second)
+				// state gets updated by watch
+				if obj.OpState != dbus.MakeVariant(OpStateNoCarrier) {
+					break
+				}
+			}
+			// error if the interface isn't connected
+			if obj.OpState == dbus.MakeVariant(OpStateNoCarrier) {
+				return false, fmt.Errorf("connection failed")
+			}
+		}
+	}
+
+	if !apply {
+		return false, nil
+	}
+
+	log.Printf("%s: ifaceCheckApply(%t)", obj, apply)
+
+	// enable or disable the interface
+	cmd := exec.Command("ip", "link", "set", obj.GetName(), obj.State)
+
+	// make sure we run the command as root
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid: true,
+		Pgid:    0,
+	}
+
+	// open a pipe to get error messages from os/exec
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return false, errwrap.Wrapf(err, "failed to initialize stderr pipe")
+	}
+
+	// start the command
+	if err := cmd.Start(); err != nil {
+		return false, errwrap.Wrapf(err, "cmd failed to start")
+	}
+	// capture any error messages
+	slurp, err := ioutil.ReadAll(stderr)
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error slurping error message")
+	}
+	// wait until cmd exits and return error message if any
+	if err := cmd.Wait(); err != nil {
+		return false, errwrap.Wrapf(err, "%s", slurp)
+	}
+
+	return false, nil
+}
+
+// fileCheckApply checks and maintains the systemd-networkd unit file contents.
+// TODO: check for conflicting unit files and think about unit priority
+func (obj *NetRes) fileCheckApply(apply bool) (checkOK bool, _ error) {
+	// check if the unit file exists
+	_, err := os.Stat(obj.UnitFilePath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return false, errwrap.Wrapf(err, "error checking file")
+		}
+	}
+	// build the unit file contents from the definition
+	contents := unitFileContents(obj.GetName(), obj.IPv4, obj.IPv6, obj.Gateway)
```

Comment:
It's probably okay to just grab these values off the struct directly and have `unitFileContents` be defined as part of the struct, so you don't have to pass them in.